File: mobile_app/lib/providers/budget_provider.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

// Модель для одной категории
class CategoryData {
  final String name;
  final double limit;
  final double spent;
  final double remaining;
  final double percent;

  CategoryData({
    required this.name,
    required this.limit,
    required this.spent,
    required this.remaining,
    required this.percent,
  });
}

class BudgetProvider extends ChangeNotifier {
  double _income = 0;
  List<CategoryData> _categories = [];
  List<String> _alerts = [];
  bool _isLoading = false;

  // Геттеры для доступа к данным извне
  double get income => _income;
  List<CategoryData> get categories => _categories;
  List<String> get alerts => _alerts;
  bool get isLoading => _isLoading;

  final String baseUrl = 'http://10.58.245.61:8080/api/budget';

  // Загрузка статуса с сервера
  Future<void> loadStatus() async {
    _isLoading = true;
    notifyListeners();
    try {
      final response = await http.get(Uri.parse('$baseUrl/status'));
      if (response.statusCode == 200) {
        final Map<String, dynamic> data = json.decode(response.body);
        _income = data['income'] ?? 0;
        _categories = [];
        if (data['categories'] != null) {
          (data['categories'] as Map<String, dynamic>).forEach((key, value) {
            _categories.add(CategoryData(
              name: key,
              limit: value['limit']?.toDouble() ?? 0,
              spent: value['spent']?.toDouble() ?? 0,
              remaining: value['remaining']?.toDouble() ?? 0,
              percent: value['percent']?.toDouble() ?? 0,
            ));
          });
        }
        // Загружаем алерты
        final alertsResponse = await http.get(Uri.parse('$baseUrl/alerts'));
        if (alertsResponse.statusCode == 200) {
          _alerts = List<String>.from(json.decode(alertsResponse.body));
        }
      }
    } catch (e) {
      print("Ошибка загрузки: $e");
    }
    _isLoading = false;
    notifyListeners();
  }

  // Отправка настроек бюджета
  Future<void> setupBudget(double income, Map<String, double> percentages) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/setup'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'income': income,
          'percentages': percentages,
        }),
      );
      print("Ответ от сервера: ${response.body}");
      await loadStatus(); // Перезагружаем данные после сохранения
    } catch (e) {
      print("Ошибка сохранения: $e");
    }
  }

  // Добавление тестовой транзакции 
  Future<void> addTestTransaction(double amount, String desc) async {
    try {
      await http.post(
        Uri.parse('$baseUrl/transaction'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'amount': amount,
          'description': desc,
        }),
      );
      await loadStatus(); // Обновляем данные после траты
    } catch (e) {
      print("Ошибка добавления транзакции: $e");
    }
  }
}

File: mobile_app/app/providers/budget.provider.ts
// providers/budget.provider.ts
import { BaseProvider } from './base.provider';
import {
    SetupBudgetRequest,
    SetupBudgetResponse,
    BudgetStatusResponse,
    BudgetAlertsResponse,
    BudgetCategoryStatus
} from '../models/budget.types';

export class BudgetProvider extends BaseProvider {
    async setupBudget(data: SetupBudgetRequest): Promise<SetupBudgetResponse> {
        try {
            if (!data.plannedIncome || data.plannedIncome <= 0) {
                throw new Error('Плановый доход должен быть больше 0');
            }

            if (!data.percentages || Object.keys(data.percentages).length === 0) {
                throw new Error('Не указаны проценты для категорий');
            }

            const totalPercent = Object.values(data.percentages).reduce((sum, p) => sum + p, 0);
            if (Math.abs(totalPercent - 100) > 0.01) {
                throw new Error(`Сумма процентов должна быть 100, а не ${totalPercent}`);
            }

            for (const [categoryId, percent] of Object.entries(data.percentages)) {
                if (percent < 0 || percent > 100) {
                    throw new Error(`Процент для категории ${categoryId} должен быть от 0 до 100`);
                }
            }

            const response = await this.request<SetupBudgetResponse>(
                '/budget/setup',
                {
                    method: 'POST',
                    body: {
                        plannedIncome: data.plannedIncome,
                        percentages: data.percentages
                    }
                }
            );
            return response;

        } catch (error) {
            if (error instanceof Error) {
                if (error.message.includes('400')) {
                    throw new Error('Неверные данные бюджета. Проверьте проценты.');
                }
                if (error.message.includes('401')) {
                    throw new Error('Не удалось сохранить бюджет. Сессия истекла.');
                }
                throw error;
            }

            this.handleError(error);
        }
    }

    async getBudgetStatus(): Promise<BudgetStatusResponse> {
        try {
            const response = await this.request<BudgetStatusResponse>(
                '/budget/status',
                { method: 'GET' }
            );
            return response;

        } catch (error) {
            if (error instanceof Error) {
                if (error.message.includes('401')) {
                    throw new Error('Не удалось загрузить статус бюджета. Сессия истекла.');
                }
                throw error;
            }

            this.handleError(error);
        }
    }

    async getAlerts(): Promise<BudgetAlertsResponse> {
        try {
            const response = await this.request<BudgetAlertsResponse>(
                '/budget/alerts',
                { method: 'GET' }
            );
            return response;

        } catch (error) {
            if (error instanceof Error) {
                if (error.message.includes('401')) {
                    throw new Error('Не удалось загрузить уведомления. Сессия истекла.');
                }
                throw error;
            }

            this.handleError(error);
        }
    }

    getRemainingForCategory(category: BudgetCategoryStatus): number {
        return category.remaining;
    }

    isCategoryOverLimit(category: BudgetCategoryStatus): boolean {
        return category.spent > category.limit;
    }

    getCategoryUsagePercent(category: BudgetCategoryStatus): number {
        return category.spentPercent;
    }

    getTotalUsagePercent(status: BudgetStatusResponse): number {
        if (status.categories.length === 0) return 0;

        const totalSpent = status.categories.reduce((sum, c) => sum + c.spent, 0);
        const totalLimit = status.categories.reduce((sum, c) => sum + c.limit, 0);

        return totalLimit > 0 ? (totalSpent / totalLimit) * 100 : 0;
    }

    sortCategoriesBySpentPercent(
        categories: BudgetCategoryStatus[],
        descending: boolean = true
    ): BudgetCategoryStatus[] {
        return [...categories].sort((a, b) =>
            descending
                ? b.spentPercent - a.spentPercent
                : a.spentPercent - b.spentPercent
        );
    }

    getHighRiskCategories(
        categories: BudgetCategoryStatus[],
        threshold: number = 80
    ): BudgetCategoryStatus[] {
        return categories.filter(c => c.spentPercent >= threshold);
    }

    getExhaustedCategories(categories: BudgetCategoryStatus[]): BudgetCategoryStatus[] {
        return categories.filter(c => c.remaining <= 0);
    }
}
